File: multi-processor-programming-study/solutions/java/Exercise01_PetersonLock_Solution.java
package com.multiprocessor.mutual_exclusion;

/**
 * SOLUTION for Peterson's Lock Exercise
 *
 * This is a reference implementation of Peterson's two-thread mutual exclusion algorithm.
 */
public class Exercise01_PetersonLock_Solution {

    public static class PetersonLock {
        // Flags indicate if a thread wants to enter critical section
        private volatile boolean[] flag = new boolean[2];

        // Victim is the thread that will defer to the other
        private volatile int victim;

        /**
         * SOLUTION: Peterson's lock protocol
         */
        public void lock(int threadId) {
            int other = 1 - threadId;

            // Step 1: Indicate interest
            flag[threadId] = true;

            // Step 2: Set yourself as victim (be polite!)
            victim = threadId;

            // Step 3: Wait while other is interested AND you are the victim
            // If you're not the victim, you can proceed
            // If other is not interested, you can proceed
            while (flag[other] && victim == threadId) {
                // Busy wait (spin)
                // This is where the waiting happens
            }
            // At this point, we have acquired the lock
        }

        /**
         * SOLUTION: Release the lock
         */
        public void unlock(int threadId) {
            // Simply indicate that we're no longer interested
            flag[threadId] = false;
        }
    }

    /**
     * Test harness
     */
    static class Counter {
        private int count = 0;

        public void increment() {
            count++;
        }

        public int getCount() {
            return count;
        }
    }

    /**
     * Demonstration
     */
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Peterson Lock Test ===");

        PetersonLock lock = new PetersonLock();
        Counter counter = new Counter();
        int incrementsPerThread = 100000;

        Thread thread0 = new Thread(() -> {
            for (int i = 0; i < incrementsPerThread; i++) {
                lock.lock(0);
                try {
                    counter.increment();
                } finally {
                    lock.unlock(0);
                }
            }
        });

        Thread thread1 = new Thread(() -> {
            for (int i = 0; i < incrementsPerThread; i++) {
                lock.lock(1);
                try {
                    counter.increment();
                } finally {
                    lock.unlock(1);
                }
            }
        });

        long startTime = System.currentTimeMillis();
        thread0.start();
        thread1.start();

        thread0.join();
        thread1.join();
        long endTime = System.currentTimeMillis();

        int expected = incrementsPerThread * 2;
        int actual = counter.getCount();

        System.out.println("Expected count: " + expected);
        System.out.println("Actual count: " + actual);
        System.out.println("Time taken: " + (endTime - startTime) + "ms");

        if (actual == expected) {
            System.out.println("✅ PASS - Mutual exclusion preserved!");
        } else {
            System.out.println("❌ FAIL - Race condition detected!");
            System.out.println("Difference: " + (expected - actual));
        }
    }
}

File: multi-processor-programming-study/solutions/cpp/exercise01_peterson_lock_solution.cpp
/**
 * SOLUTION for Peterson's Lock Exercise (C++)
 *
 * This is a reference implementation of Peterson's two-thread mutual exclusion algorithm.
 */

#include <iostream>
#include <thread>
#include <atomic>
#include <array>
#include <chrono>

namespace multiprocessor::mutual_exclusion {

/**
 * SOLUTION: Peterson's Lock implementation
 */
class PetersonLock {
private:
    std::array<std::atomic<bool>, 2> flag;
    std::atomic<int> victim;

public:
    PetersonLock() {
        flag[0].store(false, std::memory_order_relaxed);
        flag[1].store(false, std::memory_order_relaxed);
        victim.store(0, std::memory_order_relaxed);
    }

    /**
     * SOLUTION: Peterson's lock protocol
     */
    void lock(int thread_id) {
        int other = 1 - thread_id;

        // Step 1: Indicate interest
        flag[thread_id].store(true, std::memory_order_seq_cst);

        // Step 2: Set yourself as victim (be polite!)
        victim.store(thread_id, std::memory_order_seq_cst);

        // Step 3: Wait while other is interested AND you are the victim
        while (flag[other].load(std::memory_order_seq_cst) &&
               victim.load(std::memory_order_seq_cst) == thread_id) {
            // Busy wait (spin)
        }
        // At this point, we have acquired the lock
    }

    /**
     * SOLUTION: Release the lock
     */
    void unlock(int thread_id) {
        // Simply indicate that we're no longer interested
        flag[thread_id].store(false, std::memory_order_seq_cst);
    }
};

/**
 * Test harness
 */
class Counter {
private:
    int count = 0;

public:
    void increment() {
        count++;
    }

    int get_count() const {
        return count;
    }
};

} // namespace multiprocessor::mutual_exclusion

/**
 * Demonstration
 */
int main() {
    using namespace multiprocessor::mutual_exclusion;

    std::cout << "=== Peterson Lock Test ===\n";

    PetersonLock lock;
    Counter counter;
    const int increments_per_thread = 100000;

    auto thread_func = [&](int thread_id) {
        for (int i = 0; i < increments_per_thread; ++i) {
            lock.lock(thread_id);
            counter.increment();
            lock.unlock(thread_id);
        }
    };

    auto start_time = std::chrono::high_resolution_clock::now();

    std::thread thread0(thread_func, 0);
    std::thread thread1(thread_func, 1);

    thread0.join();
    thread1.join();

    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
        end_time - start_time);

    int expected = increments_per_thread * 2;
    int actual = counter.get_count();

    std::cout << "Expected count: " << expected << "\n";
    std::cout << "Actual count: " << actual << "\n";
    std::cout << "Time taken: " << duration.count() << "ms\n";

    if (actual == expected) {
        std::cout << "✅ PASS - Mutual exclusion preserved!\n";
    } else {
        std::cout << "❌ FAIL - Race condition detected!\n";
        std::cout << "Difference: " << (expected - actual) << "\n";
    }

    return 0;
}
